#AUTHOR: dwright / haerty group / earlham institute
#EMAIL: wilfried.haerty@earlham.ac.uk
#DATE: 2021 03 11
#TITLE: CACNA2D2_transcript_plotting
#DESCRIPTION: figure generation for CACNA2D2 coverage plots (short-read sequencing and GTEx cortex RNAseq samples, see methods)
#LICENCE: MIT

#Clear out R 
rm(list=ls())
#Set the working directory to my file
setwd("~/datafiles/CACNA2D2/")
getwd()

#import libraries 
library(plyr) #rename columns easily

#1 plot short-read coverage generated by coveragebed -split (see methods)
xlabel <- "Position on novel exon (minus strand 5'-3')"
indata <- read.csv2('CACNA2D2_novelexon_SHSY5Y_allSRS_coveragebed_out_itemised_split.txt', sep = '\t', header = FALSE, dec = '.')
head(indata)
colnames(indata) <- c('chr','start','end','transcriptID','strand','position','reads', 'pos_rev','reads_rev')
head(indata)
plot(indata$position, indata$reads_rev, type = 'l', ylab='Short read coverage depth', xlab=xlabel, ylim = c(200,3200))

#2 plot coveragebed -split from GTEx cortex RNAseq N=27 samples (see methods)
indata <- read.csv2('CACNA2D2_novelexon_GTEx_CORTEXonly_coveragebed_out_itemised_split.txt', sep = '\t', header = FALSE, dec = '.')
head(indata)
colnames(indata) <- c('chr','start','end','transcriptID','strand','position','reads', 'pos_rev','reads_rev')
head(indata)
plot(indata$position, indata$reads_rev, type = 'l', ylab='GTEx cortex (n=27) read coverage depth', xlab=xlabel, ylim=c(20,40))
