#AUTHOR: dwright / haerty group / earlham institute
#EMAIL: wilfried.haerty@earlham.ac.uk
#DATE: 2021 04 12
#TITLE: ONT_SRS_sequins__analysis
#DESCRIPTION: sequin spike-in assessment of Nanopore (ONT), Illumina short-read-sequencing (SRS) and downsampled Illumina (ILL)    
# LICENCE: MIT

#Clear out R 
rm(list=ls())
#Set the working directory 
setwd("~/")
getwd()

#Import libraries 
library(dplyr)
library(edgeR) #DGE package
library(Anaquin) # Sequins package
library(gridExtra)
library(matrixStats)
library(reshape2)
library(ggplot2)
library(gghighlight)
library(ggrepel)

# Note: abbreviations: ONT = nanopore, SRS = full Illumina "Short-Read_Sequencing", ILL = downsampled Illumina (see methods)

### import sequin details for both transcript and gene level
sequins_list <- read.csv('~/datafiles/SEQUINS/sequin_files/sequins_transcripts_concs.txt', sep = '\t')
sequins_gene <- read.csv('~/datafiles/SEQUINS/sequin_files/sequin_genes_concs.txt', sep = '\t')

### import ONT SRS and ILL TPM data generated by salmon (transcript-level quantification, 'tran')
ONTtran <- read.csv2('~/datafiles/SEQUINS/tpm_files/ONT_transcripts_TPM_sequins.txt', sep = '\t', header = TRUE, dec = '.')
SRStran <- read.csv2('~/datafiles/SEQUINS/tpm_files/SRS_remapped_transcriptome_sequins_isoform_TPM.txt', sep = '\t', header = TRUE, dec = '.' )
ILLtran <- read.csv2('~/datafiles/SEQUINS/tpm_files/downsampled_ILL_sequins_transcript_TPM.txt', sep = '\t', header = TRUE, dec = '.')

## make gene-level TPM from transcript-level TPM using script from I.Sudbery https://www.biostars.org/p/329625/
# quotation:"As TPM is transcripts per million, the gene TPM is simply the some of the transcript TPMs for all transcripts belonging to that gene. If our dataframe transcript_tpm has gene_id, transcript_id and TPM, then we calculate gene TPM using dplyr thus: gene_tpm <- group_by(transcript_tpm, gene_id) %>% summarize(TPM = sum(TPM))"
ONTgene <- group_by(ONTtran, gene_id) %>% summarize(D1=sum(D1), D2=sum(D2), D3=sum(D3), D4=sum(D4), D5=sum(D5), U1=sum(U1), U2=sum(U2), U3=sum(U3), U4=sum(U4), U5=sum(U5))
SRSgene <- group_by(SRStran, gene_id) %>% summarize(U1=sum(U1), U2=sum(U2), U3=sum(U3), U4=sum(U4), U5=sum(U5), D1=sum(D1), D2=sum(D2), D3=sum(D3), D4=sum(D4), D5=sum(D5))
ILLgene <- group_by(ILLtran, gene_id) %>% summarize(U1_downsampled=sum(U1_downsampled), U2_downsampled=sum(U2_downsampled), U3_downsampled=sum(U3_downsampled), U4_downsampled=sum(U4_downsampled), U5_downsampled=sum(U5_downsampled), D1_downsampled=sum(D1_downsampled), D2_downsampled=sum(D2_downsampled), D3_downsampled=sum(D3_downsampled), D4_downsampled=sum(D4_downsampled), D5_downsampled=sum(D5_downsampled))

############################################################################################################
############################################################################################################
###### Sequencing reads vs sequin concentration figures (both technologies + downsampled Illumina) 
############################################################################################################
############################################################################################################

################################
#### ONT 
################################
#####
# transcript-level
#####

# split ONTtran into mixA and mixB and order by sequinID
ONTtranB <- ONTtran[,c(1,2,3,5,7,9,11)]
ONTtranB <- ONTtranB[order(ONTtranB$transcript_id),]
ONTtranA <- ONTtran[,c(1,2,4,6,8,10,12)]
ONTtranA <- ONTtranA[order(ONTtranA$transcript_id),]

# # plot mixA
ONTreadsA <- log2(ONTtranA[,3:7])
is.na(ONTreadsA)<-sapply(ONTreadsA, is.infinite)
ONTreadsA[is.na(ONTreadsA)]<-0
ONTnamesA <- sequins_list$ID
ONTinputA <- log2(sequins_list$V2_MixA) # MixA!
is.na(ONTinputA)<-sapply(ONTinputA, is.infinite)
ONTinputA[is.na(ONTinputA)]<-0
ONTtitleA <- 'ONT isoform expression MixA'
xlab  <- expression(Input~concentration~(log[2]))
ylab  <- expression(Measured~TPM~(log[2]))
plotLinear(ONTnamesA, ONTinputA, ONTreadsA, title="", xlab=xlab, ylab=ylab, showLOQ = TRUE)
ONTreadsAnames <- cbind(ONTreadsA, ONTtranA$transcript_id)
colnames(ONTreadsAnames)[which(names(ONTreadsAnames) == "ONTtranA$transcript_id")] <- "name"

# plot mixB
ONTreadsB <- log2(ONTtranB[,3:7])
is.na(ONTreadsB)<-sapply(ONTreadsB, is.infinite)
ONTreadsB[is.na(ONTreadsB)]<-0
ONTnamesB <- sequins_list$ID
ONTinputB <- log2(sequins_list$V1_MixB) # MixB!
is.na(ONTinputB)<-sapply(ONTinputB, is.infinite)
ONTinputB[is.na(ONTinputB)]<-0
title <- 'ONT Isoform expression Mix B'
xlab  <- expression(Input~concentration~(log[2]))
ylab  <- expression(Measured~TPM~(log[2]))
plotLinear(ONTnamesB, ONTinputB, ONTreadsB, title="", xlab=xlab, ylab=ylab, showLOQ = TRUE)
ONTreadsBnames <- cbind(ONTreadsB, ONTtranB$transcript_id)
colnames(ONTreadsBnames)[which(names(ONTreadsBnames) == "ONTtranB$transcript_id")] <- "name"

#####
# gene -level
#####

# split ONTgene into mixA and mixB and order by sequinID
ONTgeneA <- ONTgene[,c(1,3,5,7,9,11)]
ONTgeneA <- ONTgeneA[order(ONTgeneA$gene_id),]
ONTgeneB <- ONTgene[,c(1,2,4,6,8,10)]
ONTgeneB <- ONTgeneB[order(ONTgeneB$gene_id),]

# plot mixA
ONTgA <- log2(ONTgeneA[,2:6])
is.na(ONTgA)<-sapply(ONTgA, is.infinite)
ONTgA[is.na(ONTgA)]<-0
ONTnamesG <- sequins_gene$gene
ONTinputG_A <- log2(sequins_gene$V2_MixA) # MixA!
is.na(ONTinputG_A)<-sapply(ONTinputG_A, is.infinite)
ONTinputG_A[is.na(ONTinputG_A)]<-0
ONTtitlegA <- ''
xlab  <- expression(Input~concentration~(log[2]))
ylab  <- expression(Measured~TPM~(log[2]))
plotLinear(ONTnamesG, ONTinputG_A, ONTgA, title=ONTtitlegA, xlab=xlab, ylab=ylab, showLOQ = TRUE)
ONTgAnames <- cbind(ONTgA, ONTgeneA$gene_id)
colnames(ONTgAnames)[which(names(ONTgAnames) == "ONTgeneA$gene_id")] <- "name"

# plot mixB
ONTgB <- log2(ONTgeneB[,2:6])
is.na(ONTgB)<-sapply(ONTgB, is.infinite)
ONTgB[is.na(ONTgB)]<-0
ONTnamesG <- sequins_gene$gene
ONTinputG_B <- log2(sequins_gene$V1_MixB) # MixB!
is.na(ONTinputG_B)<-sapply(ONTinputG_B, is.infinite)
ONTinputG_B[is.na(ONTinputG_B)]<-0
ONTtitlegB <- ''
xlab  <- expression(Input~concentration~(log[2]))
ylab  <- expression(Measured~TPM~(log[2]))
plotLinear(ONTnamesG, ONTinputG_B, ONTgB, title=ONTtitlegB, xlab=xlab, ylab=ylab, showLOQ = TRUE)
ONTgBnames <- cbind(ONTgB, ONTgeneB$gene_id)
colnames(ONTgBnames)[which(names(ONTgBnames) == "ONTgeneB$gene_id")] <- "name"

################################
#### SRS (Full short-read sequencing dataset)
################################

#####
# transcript level
#####

# split SRStran into mixA and mixB
SRStranB <- SRStran[,c(1,2,4,6,8,10,12)]
SRStranB <- SRStranB[order(SRStranB$NAME),]
SRStranA <- SRStran[,c(1,2,3,5,7,9,11)]
SRStranA <- SRStranA[order(SRStranA$NAME),]

# plot mixA
readsSRSA <- log2(SRStranA[,3:7])
is.na(readsSRSA)<-sapply(readsSRSA, is.infinite)
readsSRSA[is.na(readsSRSA)]<-0
SRSnamesA <- sequins_list$ID
SRSinputA <- log2(sequins_list$V2_MixA) # MixA!
is.na(SRSinputA)<-sapply(SRSinputA, is.infinite)
SRSinputA[is.na(SRSinputA)]<-0
SRStitleA <- ''
xlab  <- expression(Input~concentration~(log[2]))
ylab  <- expression(Measured~TPM~(log[2]))
plotLinear(SRSnamesA, SRSinputA, readsSRSA, title=SRStitleA, xlab=xlab, ylab=ylab, showLOQ = TRUE)
readsSRSAnames <- cbind(readsSRSA, SRStranA$NAME)
colnames(readsSRSAnames)[which(names(readsSRSAnames) == "SRStranA$NAME")] <- "name"

# plot mixB
readsSRSB <- log2(SRStranB[,3:7])
is.na(readsSRSB)<-sapply(readsSRSB, is.infinite)
readsSRSB[is.na(readsSRSB)]<-0
namesB <- sequins_list$ID
inputB <- log2(sequins_list$V1_MixB) # MixB!
is.na(inputB)<-sapply(inputB, is.infinite)
inputB[is.na(inputB)]<-0
title <- ''
xlab  <- expression(Input~concentration~(log[2]))
ylab  <- expression(Measured~TPM~(log[2]))
plotLinear(namesB, inputB, readsSRSB, title=title, xlab=xlab, ylab=ylab, showLOQ = TRUE)
readsSRSBnames <- cbind(readsSRSB, SRStranB$NAME)
colnames(readsSRSBnames)[which(names(readsSRSBnames) == "SRStranB$NAME")] <- "name"

#####
# gene level
#####

# split SRSgene into mixA and mixB and order by sequinID
SRSgeneB <- SRSgene[,c(1,3,5,7,9,11)]
SRSgeneB <- SRSgeneB[order(SRSgeneB$gene_id),]
SRSgeneA <- SRSgene[,c(1,2,4,6,8,10)]
SRSgeneA <- SRSgeneA[order(SRSgeneA$gene_id),]

# plot mixA
SRSgA <- log2(SRSgeneA[,2:6])
is.na(SRSgA)<-sapply(SRSgA, is.infinite)
SRSgA[is.na(SRSgA)]<-0
SRSnamesG <- sequins_gene$gene
SRSinputG_A <- log2(sequins_gene$V2_MixA) # MixA!
is.na(SRSinputG_A)<-sapply(SRSinputG_A, is.infinite)
SRSinputG_A[is.na(SRSinputG_A)]<-0
SRStitlegA <- ''
xlab  <- expression(Input~concentration~(log[2]))
ylab  <- expression(Measured~TPM~(log[2]))
plotLinear(SRSnamesG, SRSinputG_A, SRSgA, title=SRStitlegA, xlab=xlab, ylab=ylab, showLOQ = TRUE)
SRSgAnames <- cbind(SRSgA, SRSgeneA$gene_id)
colnames(SRSgAnames)[which(names(SRSgAnames) == "SRSgeneA$gene_id")] <- "name"

# plot mixB
SRSgB <- log2(SRSgeneB[,2:6])
is.na(SRSgB)<-sapply(SRSgB, is.infinite)
SRSgB[is.na(SRSgB)]<-0
SRSnamesG <- sequins_gene$gene
SRSinputG_B <- log2(sequins_gene$V1_MixB) # MixB!
is.na(SRSinputG_B)<-sapply(SRSinputG_B, is.infinite)
SRSinputG_B[is.na(SRSinputG_B)]<-0
SRStitlegB <- ''
xlab  <- expression(Input~concentration~(log[2]))
ylab  <- expression(Measured~TPM~(log[2]))
plotLinear(SRSnamesG, SRSinputG_B, SRSgB, title=SRStitlegB, xlab=xlab, ylab=ylab, showLOQ = TRUE)
SRSgBnames <- cbind(SRSgB, SRSgeneB$gene_id)
colnames(SRSgBnames)[which(names(SRSgBnames) == "SRSgeneB$gene_id")] <- "name"

################################
#### ILLUMINA (downsampled) 
################################

#####
# transcript level
#####

# split ILLtran into mixA and mixB. 
ILLtranB <- ILLtran[,c(1,2,4,6,8,10,12)]
ILLtranB <- ILLtranB[order(ILLtranB$transcript_id),]
ILLtranA <- ILLtran[,c(1,2,3,5,7,9,11)]
ILLtranA <- ILLtranA[order(ILLtranA$transcript_id),]

# plot mixA
readsILLA <- log2(ILLtranA[,3:7])
is.na(readsILLA)<-sapply(readsILLA, is.infinite)
readsILLA[is.na(readsILLA)]<-0
IllnamesA <- sequins_list$ID
IllinputA <- log2(sequins_list$V2_MixA) # MixA!
is.na(IllinputA)<-sapply(IllinputA, is.infinite)
IllinputA[is.na(IllinputA)]<-0
IlltitleA <- ''
xlab  <- expression(Input~concentration~(log[2]))
ylab  <- expression(Measured~TPM~(log[2]))
plotLinear(IllnamesA, IllinputA, readsILLA, title=IlltitleA, xlab=xlab, ylab=ylab, showLOQ = TRUE)
readsILLAnames <- cbind(readsILLA, ILLtranA$transcript_id)
colnames(readsILLAnames)[which(names(readsILLAnames) == "ILLtranA$transcript_id")] <- "name"

# plot mixB
readsILLB <- log2(ILLtranB[,3:7])
is.na(readsILLB)<-sapply(readsILLB, is.infinite)
readsILLB[is.na(readsILLB)]<-0
namesB <- sequins_list$ID
inputB <- log2(sequins_list$V1_MixB) # MixB!
is.na(inputB)<-sapply(inputB, is.infinite)
inputB[is.na(inputB)]<-0
title <- ''
xlab  <- expression(Input~concentration~(log[2]))
ylab  <- expression(Measured~TPM~(log[2]))
plotLinear(namesB, inputB, readsILLB, title=title, xlab=xlab, ylab=ylab, showLOQ = TRUE)
readsILLBnames <- cbind(readsILLB, ILLtranB$transcript_id)
colnames(readsILLBnames)[which(names(readsILLBnames) == "ILLtranB$transcript_id")] <- "name"

#####
# Gene Level
#####

# split Illgene into mixA and mixB and order by sequinID
ILLgeneB <- ILLgene[,c(1,3,5,7,9,11)]
ILLgeneB <- ILLgeneB[order(ILLgeneB$gene_id),]
ILLgeneA <- ILLgene[,c(1,2,4,6,8,10)]
ILLgeneA <- ILLgeneA[order(ILLgeneA$gene_id),]

# plot mixA
ILLgA <- log2(ILLgeneA[,2:6])
is.na(ILLgA)<-sapply(ILLgA, is.infinite)
ILLgA[is.na(ILLgA)]<-0
ILLnamesG <- sequins_gene$gene
ILLinputG_A <- log2(sequins_gene$V2_MixA) # MixA!
is.na(ILLinputG_A)<-sapply(ILLinputG_A, is.infinite)
ILLinputG_A[is.na(ILLinputG_A)]<-0
ILLtitlegA <- ''
xlab  <- expression(Input~concentration~(log[2]))
ylab  <- expression(Measured~TPM~(log[2]))
plotLinear(ILLnamesG, ILLinputG_A, ILLgA, title=ILLtitlegA, xlab=xlab, ylab=ylab, showLOQ = TRUE)
ILLgAnames <- cbind(ILLgA, ILLgeneA$gene_id)
colnames(ILLgAnames)[which(names(ILLgAnames) == "ILLgeneA$gene_id")] <- "name"

# plot mixB
ILLgB <- log2(ILLgeneB[,2:6])
is.na(ILLgB)<-sapply(ILLgB, is.infinite)
ILLgB[is.na(ILLgB)]<-0
ILLnamesG <- sequins_gene$gene
ILLinputG_B <- log2(sequins_gene$V1_MixB) # MixB!
is.na(ILLinputG_B)<-sapply(ILLinputG_B, is.infinite)
ILLinputG_B[is.na(ILLinputG_B)]<-0
ILLtitlegB <- ''
xlab  <- expression(Input~concentration~(log[2]))
ylab  <- expression(Measured~TPM~(log[2]))
plotLinear(ILLnamesG, ILLinputG_B, ILLgB, title=ILLtitlegB, xlab=xlab, ylab=ylab, showLOQ = TRUE)
ILLgBnames <- cbind(ILLgB, ILLgeneB$gene_id)
colnames(ILLgBnames)[which(names(ILLgBnames) == "ILLgeneB$gene_id")] <- "name"

############################################################################################################
############################################################################################################
## COMPARE ILLUMINA AND ONT EXPRESSION FIGURES (ISOFORM LEVEL)
############################################################################################################
############################################################################################################

###############################
### MixA isoform level between ONT and full Illumina short-read sequencing dataset (SRS) 
###############################

xdataIA <- ONTreadsAnames %>% mutate(Mean= rowMeans(.[1:5]), stdev=rowSds(as.matrix(.[1:5])))
names(xdataIA)[7] <- 'Xmean' 
names(xdataIA)[8] <- 'Xsd'
ydataIA <- readsSRSAnames %>% mutate(Mean= rowMeans(.[1:5]), stdev=rowSds(as.matrix(.[1:5])))
names(ydataIA)[7] <- 'Ymean'
names(ydataIA)[8] <- 'Ysd'
compdataIA <- cbind.data.frame(xdataIA$Xmean, xdataIA$Xsd, ydataIA$Ymean, ydataIA$Ysd, xdataIA$name)
colnames(compdataIA) <- c('Xmean', 'Xsd', 'Ymean', 'Ysd','name')
compdataIA <- data.frame(compdataIA)
compdataIA$grp <- as.factor(abs(compdataIA$Xmean))
compdataIA$ymax <- compdataIA$Ymean + 2*compdataIA$Ysd # calculate the 95% CI by add/subtract 2 SDs from mean
compdataIA$ymin <- compdataIA$Ymean - 2*compdataIA$Ysd
compdataIA$xmax <- compdataIA$Xmean + 2*compdataIA$Xsd
compdataIA$xmin <- compdataIA$Xmean - 2*compdataIA$Xsd
compdataIA

# prepare plot labels
compIA <- 'MixA isoform expression ONT vs SRS'
xlab  <- expression(ONT~measured~TPM~(log[2]))
ylab  <- expression(SRS~measured~TPM~(log[2]))
threshIA <- subset(compdataIA,compdataIA$name == 'R1_52_1') # identify this manually from sequin concentration table & earlier plots showing LOQ for ONT isoform mix

# plot for paneling Fig1A: isoform-level ONT vs SRS sequins of mixA
highlight_point <- subset(compdataIA, compdataIA$name=='R1_52_1')
plot1 <- ggplot(data = compdataIA, aes(x=Ymean, y=Xmean)) + coord_cartesian(ylim = c(-12, 15)) + xlab(ylab) + ylab(xlab) + theme_bw(base_size = 14) + labs(colour='Ratio') + geom_errorbar(aes(ymin = xmin,ymax = xmax), alpha = 0.4) + geom_errorbarh(aes(xmin = ymin,xmax = ymax), alpha = 0.3) + geom_point(colour='grey13',size=2.4, alpha=0.9) + guides(colour=FALSE) + ggtitle('') + geom_label_repel(data = threshIA, aes(x=Ymean, y=Xmean), label = 'R1_52_1', label.padding = unit(0.20, 'lines'), label.size = 0.4, colour = 'black', min.segment.length = unit(0,'lines'),  nudge_x = 3, nudge_y = -3) + annotate(geom = 'text', size = 12, x =14 , y = -10, label = 'A') + geom_point(data = highlight_point, aes(x=Ymean, y=Xmean), color='red', size=2.4, alpha=0.9)


###############################
### MixB isoform level between ONT and full Illumina short-read sequencing dataset (SRS) 
###############################

xdataIB <- ONTreadsBnames %>% mutate(Mean= rowMeans(.[1:5]), stdev=rowSds(as.matrix(.[1:5])))
names(xdataIB)[7] <- 'Xmean' 
names(xdataIB)[8] <- 'Xsd'
ydataIB <- readsSRSBnames %>% mutate(Mean= rowMeans(.[1:5]), stdev=rowSds(as.matrix(.[1:5])))
names(ydataIB)[7] <- 'Ymean'
names(ydataIB)[8] <- 'Ysd'
compdataIB <- cbind.data.frame(xdataIB$Xmean, xdataIB$Xsd, ydataIB$Ymean, ydataIB$Ysd, xdataIB$name)
colnames(compdataIB) <- c('Xmean', 'Xsd', 'Ymean', 'Ysd','name')
compdataIB <- data.frame(compdataIB)
compdataIB$grp <- as.factor(abs(compdataIB$Xmean))
compdataIB$ymax <- compdataIB$Ymean + 2*compdataIB$Ysd # calculate the 95% CI by add/subtract 2 SDs from mean
compdataIB$ymin <- compdataIB$Ymean - 2*compdataIB$Ysd
compdataIB$xmax <- compdataIB$Xmean + 2*compdataIB$Xsd
compdataIB$xmin <- compdataIB$Xmean - 2*compdataIB$Xsd
compdataIB
## plot but reverse axes to make more readable 
xlab  <- expression(ONT~measured~TPM~(log[2]))
ylab  <- expression(SRS~measured~TPM~(log[2]))
threshIB <- subset(compdataIB,compdataIB$name == 'R1_62_1') 

# plot for paneling Fig1B: isoform-level ONT vs SRS sequins of mixB
highlight_point2 <- subset(compdataIB, name == "R1_62_1")
plot2 <- ggplot(data = compdataIB, aes(x=Ymean, y=Xmean)) + coord_cartesian(ylim = c(-12, 15)) + xlab(ylab) + ylab(xlab) + theme_bw(base_size = 14) + theme(plot.title = element_text(face = 'bold')) + theme(plot.title = element_text(hjust = 0.5)) + labs(colour='Ratio') + geom_errorbar(aes(ymin = xmin,ymax = xmax), alpha = 0.4) + geom_errorbarh(aes(xmin = ymin,xmax = ymax), alpha = 0.3) + geom_point(colour='grey13',size=2.4, alpha=0.9) + guides(colour=FALSE) + ggtitle('') + geom_label_repel(data = threshIB, aes(x=Ymean, y=Xmean), label = 'R1_62_1', label.padding = unit(0.25, 'lines'), label.size = 0.4, colour = 'black', min.segment.length = unit(0,'lines'),  nudge_x = 0.5, nudge_y = 8) + annotate(geom = 'text', size = 12, x =14 , y = -10, label = 'B') + geom_point(data = highlight_point2, aes(x=Ymean, y=Xmean), color='red', size=2.4, alpha=0.9)

############################################################################################################
############################################################################################################
####  Differential Expression analysis of the Sequins at isoform level using EdgeR ############################################################################################################
############################################################################################################

# ONT Transcript counts (ONTTC)
ONTTC <- read.table('~/datafiles/SEQUINS/ONT_numReads_sequins_transcripts_Mar2021.txt', sep = '\t', header = TRUE, dec = '.')
ONTTC <- ONTTC[order(ONTTC$transcript),]

###### ONT SEQUIN TRANSCRIPT LEVEL DIFF EXPRESSION
head(ONTTC)
thresh <- 0
ONTTC_filt <- ONTTC[rowSums(ONTTC [,-1]) > thresh,]
ONTTC_filt_dgList <- DGEList(counts = as.matrix(ONTTC_filt[,2:11]), genes = ONTTC_filt$transcript, group=rep(1:2,5)) # note arrangement of groupings important here (ABABABABABA)
ONTTC_filt_dgList <- calcNormFactors(ONTTC_filt_dgList, method="TMM")
plotMDS(ONTTC_filt_dgList)
#pdf('ONT_Transcript_MDSplot.pdf', width = 10, height = 6)
#plotMDS(ONTTC_filt_dgList)
#dev.off()
eff.lib.size <- ONTTC_filt_dgList$samples$lib.size*ONTTC_filt_dgList$samples$norm.factors
normCounts <- cpm(ONTTC_filt_dgList)
pseudoNormCounts <- log2(normCounts + 1)
boxplot(pseudoNormCounts, col="gray", las=3, ylab = 'Pseudonormalised Counts', main = 'ONT transcript-level')
#pdf('ONT_Transcript_pseudonorm_counts.pdf', width = 8, height = 6)
#boxplot(pseudoNormCounts, col="gray", las=3, ylab = 'Pseudonormalised Counts', main = 'ONT transcript-level')
#dev.off()
sampleType <- c('A','B','A','B','A','B','A','B','A','B')
sampleReplicate <- paste("S", rep(1:5, 2), sep="")
sampleReplicate
designMat <- model.matrix(~sampleReplicate + sampleType)
designMat
ONTTC_filt_dgList <- estimateGLMCommonDisp(ONTTC_filt_dgList, design=designMat)
ONTTC_filt_dgList <- estimateGLMTrendedDisp(ONTTC_filt_dgList, design=designMat)
ONTTC_filt_dgList <- estimateGLMTagwiseDisp(ONTTC_filt_dgList, design=designMat)
ONTTC_filt_dgList
plotBCV(ONTTC_filt_dgList, main = 'ONT transcript-level')
#pdf('ONT_Transcript_BCV.pdf', width = 8, height = 6)
#plotBCV(ONTTC_filt_dgList, main = 'ONT transcript-level')
#dev.off()
fit <- glmFit(ONTTC_filt_dgList, designMat)
colnames(fit)
ONT_T_lrt <- glmLRT(fit, coef='sampleTypeB') # remember to check the col name of the variable (this case MixA/MixB)
edgeR_result <- topTags(ONT_T_lrt)
edgeR_result
deTranscripts <- decideTestsDGE(ONT_T_lrt, p=0.001)
deTranscripts <- rownames(ONT_T_lrt)[as.logical(deTranscripts)]
plotSmear(ONT_T_lrt, de.tags=deTranscripts, main = 'ONT transcript-level')
abline(h=c(-1, 1), col=2)
#pdf('ONT_Transcript_DEsmear.pdf', width = 8, height = 6)
#plotSmear(ONT_T_lrt, de.tags=deTranscripts, main = 'ONT transcript-level')
#abline(h=c(-1, 1), col=2)
#dev.off()
ONT_T_lrt$table$logFC # list of log Fold-change from analysis
ONT_T_lrt$table$PValue # list of Pvalues from analysis 
ONT_T_lrt$table$FDR <- p.adjust(ONT_T_lrt$tabl$PValue, method = 'fdr') # manually correct for multiple testing 
volcanoData <- cbind(ONT_T_lrt$table$logFC, -log10(ONT_T_lrt$table$FDR))
colnames(volcanoData) <- c("logFC", "negLogPval")
head(volcanoData)
plot(volcanoData, pch=19, main = 'ONT transcript-level')
#pdf('ONT_Transcript_DEvolcano.pdf', width = 8, height = 6)
#plot(volcanoData, pch=19, main = 'ONT transcript-level')
#dev.off()

############################################################################################################
############################################################################################################
## Compare log2(+1) expected FC against observed FC for each Sequin mix at isoform level (CORRELATION PLOT Fig1C)
############################################################################################################
############################################################################################################

# input manually calculated log2FC for sequins at each level (calculated by log2(mixB/mixA-1)+1) (Obs logFC exported for archiving, so re-imported with Exp logFC as single file for ease)
isoformEXPFC <- read.csv2('~/datafiles/SEQUINS/sequins_transcripts_EXP_Log2FCplus1_20210309_final.txt', sep = '\t', header = TRUE, dec = '.')
length(isoformEXPFC$transcript)

# Compare observed and expected log2FC+1 + produce plot for manuscript
cor.test(isoformEXPFC$EXP_log2FC.1, isoformEXPFC$OBS_log2FC.1)
label1 <- paste("R^2 ==", 0.973)
label2 <- paste("p == 2.2e-16")

#PLOT 3 of correlation
plot3 <- ggplot(data=isoformEXPFC, aes(x=EXP_log2FC.1, y=OBS_log2FC.1)) + xlab(expression(ONT~Expected~(log[2])~fold~change)) + ylab(expression(ONT~Observed~(log[2])~fold~change)) + theme_bw(base_size = 14) + geom_point(size = 2.5, alpha = 0.9, colour='grey16') + geom_smooth(method = lm, fill = 'grey20', color = 'grey9') + annotate("text", label = label1, parse=TRUE, x=-8, y=16, size = 5, hjust = 0) + annotate("text", label = label2, parse=TRUE, x=-8, y=12.5, size = 5, hjust = 0) + annotate(geom = 'text', size = 12, x =12 , y = -13, label = 'C') + theme(plot.margin = unit(c(1.05, 0.2, 0.25, 0.2), "cm")) #+ theme(axis.title.x = element_text(vjust=-0.185))

### PANELLING FOR FIGURE 1
# GRID ARRANGE FOR MANUSCRIPT
grid.arrange(plot1, plot2,plot3, ncol=3)